app/controllers/settings/external_cert_controller.rb

```diff
@@ -355,10 +357,56 @@ def warning_subjectaltname(cert, warning_messages, required_san_array, service_n
   end
   # rubocop:enable Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity
 
-  # # Placeholder for trust chain validation
-  # def trust_chain_verify(_cert)
-  #   true
-  # end
+  def trust_chain_verify(cert)
+    _trust_chain_verify(cert, _build_system_wide_certs_store)
+  end
+
+  def _build_system_wide_certs_store
+    cert_store = OpenSSL::X509::Store.new
+
+    system_certs = SystemCertificate.all
+    system_certs.each do |system_cert|
+      cert = system_cert.certificate
+      cert_text = cert.certificate
+      cert_obj = read_cert(cert_text)
+      # Certs must pass parsing to get into the SystemCertificate table; so we ignore failing parse
+      next unless cert_obj
+      cert_store.add_cert(cert_obj)
+    end
+
+    cert_store
+  end
+
+  # rubocop:disable Lint/RescueException
+  def _trust_chain_verify(cert_obj, cert_store)
+    verify_error = ""
+
+    # setup callback
+    cert_store.verify_callback = proc do |preverify_ok, ssl_context|
+      begin
+        if preverify_ok != true || ssl_context.error != 0
```

Comment:
This `||` should probably be `or`.  I _think_ they both have lower precedence than comparison, but it always makes me nervous to see `||` near `=` :)

Reply:
The difference is only that "or" short circuits. || does not short circuit. I will make the change; short circuiting does make sense here.